pkg/processor/controlcommunication/controlcommunication.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2023 The Nuclio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controlcommunication
+
+import (
+	"bufio"
+	"fmt"
+	"sync"
+
+	"github.com/nuclio/errors"
+)
+
+type ControlConsumer struct {
+	channels []chan *ControlMessage
+	kind     ControlMessageKind
+	lock     sync.Mutex
+}
+
+// NewControlConsumer creates a new control consumer
+func NewControlConsumer(kind ControlMessageKind) *ControlConsumer {
+
+	return &ControlConsumer{
+		channels: make([]chan *ControlMessage, 0),
+		kind:     kind,
+		lock:     sync.Mutex{},
+	}
+}
+
+// GetKind returns the kind of the consumer
+func (c *ControlConsumer) GetKind() ControlMessageKind {
+	return c.kind
+}
+
+// BroadcastAndCloseSubscriptions sends a message to all subscribed channels and deletes all subscriptions after
+func (c *ControlConsumer) BroadcastAndCloseSubscriptions(message *ControlMessage) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// do nothing if channels is an empty slice
+	if len(c.channels) == 0 {
+		return nil
+	}
+	// write message to the all channels
+	for _, channel := range c.channels {
+		go func(channel chan *ControlMessage, message *ControlMessage) {
+			defer func() {
+				// if the channel is closed before the message is read, it will result in a panic
+				if err := recover(); err != nil {
+					fmt.Println("Recovered in BroadcastAndCloseSubscriptions", err)
+				}
+			}()
+			channel <- message
+		}(channel, message)
+	}
+
+	// delete the channel from subscription
+	c.channels = make([]chan *ControlMessage, 0)
+
+	return nil
+}
+
+// Broadcast broadcasts a message to all subscribed channels
+func (c *ControlConsumer) Broadcast(message *ControlMessage) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(c.channels))
+	for _, channel := range c.channels {
+
+		go func(channel chan *ControlMessage, message *ControlMessage) {
+			defer func() {
+				// if the channel is closed before the message is read, it will result in a panic
+				if err := recover(); err != nil {
+					fmt.Println("Recovered in BroadcastAndCloseSubscriptions", err)
```

Comment:
```suggestion
					fmt.Println("Recovered in Broadcast", err)
```

I don't really like that we use `fmt.Println` here, instead of logging. We use this in very specific places.

I know we don't have a logger instance in the consumer, but this might result in uncoventional messages hidden in-between our logs.
I think we should either initialize the message broker with a logger or pass a logger instance to these functions.
(Same for the recover in `BroadcastAndCloseSubscriptions`)

Then, we'll also be able to log the error properly - `"err", err.Error()`

Reply:
Agreed, will add logging